Add User#getFlags

Andre601 opened the pull request:
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

<!--
  There are several guidelines you should follow in order for your
  Pull Request to be merged.
-->

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

<!--
  It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.
-->

### Changes

- [ ] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ <!-- Insert other type here -->

<!-- Replace "NaN" with an issue number if this is a response to an issue -->

Closes Issue: NaN

## Description

This implements the `public_flags` value for the User object, which essentially contains all [user flags](https://discord.com/developers/docs/resources/user#user-object-user-flags) with some exceptions like what type of nitro (or if they even have nitro) the user has and some other hidden flags.

duncte123 commented:
Why didn't you add the early supporter one?

Andre601 commented:
> Why didn't you add the early supporter one?

I'm not sure if all flags are available, because people said that some aren't insider the public_flags, and my assumption is, that any kind of nitro indicator (including ea) are not in this...
There's sadly no proper documentation (yet) about which flags aren't included, because if public_flags = flags, then why not just use flags?

Andre601 commented:
@duncte123 I now added all flags that are considered "public" (based on [this file](https://github.com/vDelite/DiscordLists/blob/master/flags.md))

I also renamed some flags to make them match the internal names Discord uses, although I think having the names of the HypeSquad houses would be better.

boomboompower commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +225,70 @@ default String getEffectiveAvatarUrl()
+        Flags(int offset, @Nonnull String name)
+        {
+            this.offset = offset;
+            this.raw = 1 << offset;
```
This value is incorrect, you are already passing the raw offset as the first paramater;
Example: the "raw" varaible for NONE will be 1, 2 for STAFF, 4 fpr HYPESQUAD.

To fix this issue you should change each Enums value to the offset instead of the raw bit shifted value

boomboompower commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +225,70 @@ default String getEffectiveAvatarUrl()
+        Flags(int offset, @Nonnull String name)
+        {
+            this.offset = offset;
+            this.raw = 1 << offset;
```
This value here is incorrect, you are already passing the raw offset as the first paramater;
Example: the "raw" varaible for NONE will be 1, 2 for STAFF, 4 fpr HYPESQUAD.

To fix this issue you should change each Enums value to the offset instead of the raw bit shifted value

boomboompower replied:
I've fixed it in a PR [here](https://github.com/Andre601/JDA/pull/3)

The "Flags" enum should probably be renamed to UserFlags as well

HydroPage90 replied:
@boomboompower Agreed. UserFlag would be more readable

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -230,16 +230,23 @@ default String getEffectiveAvatarUrl()
+        // Brilliance
+        HYPESQUAD_ONLINE_HOUSE_2(128, "HypeSquad Online House Brilliance"),
+        // Balance
+        HYPESQUAD_ONLINE_HOUSE_3(256, "HypeSquad Online House Balance"),
```
Let's call them `HYPESQUAD_BRAVERY` and such.

Andre601 replied:
Was renamed.
https://github.com/DV8FromTheWorld/JDA/pull/3

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -230,16 +230,23 @@ default String getEffectiveAvatarUrl()
-        BALANCE(       256, "HypeSquad House Balance"),
-        SYSTEM(       4096, "System"),
-        BUG_HUNTER_2(16384, "Bug Hunter Level 2");
+        NONE(                      0, "None"),
```
How would this be used?

Andre601 replied:
Renamed it to `UNKNOWN` with -1 as value (Placeholder).

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -230,16 +230,23 @@ default String getEffectiveAvatarUrl()
+        SYSTEM(                 4096, "System User"),
+        BUG_HUNTER_LEVEL_2(    16384, "Bug Hunter Level 2"),
+        VERIFIED_BOT(          65536, "Verified Bot"),
+        VERIFIED_DEVELOPER(   131072, "Verified Developer");
```
Instead of using raw ints, please use the offsets and calculate the raw value in the constructor.

Andre601 replied:
Was changed int https://github.com/DV8FromTheWorld/JDA/pull/3

Andre601 commented:
@MinnDevelopment How should we treat `NONE` then when using the offset?
Because `NONE` is a flag in Discord, using the value 0 and I think using `-1` should be reserved for a `UNKNOWN` flag.

I suggested [here](https://github.com/Andre601/JDA/pull/3) that we perhaps should have both the offset and the raw value? Would kind of make sense considering Discord provides the flags as their raw (total) value.

MinnDevelopment commented:
> @MinnDevelopment How should we treat NONE then when using the offset?
Because NONE is a flag in Discord, using the value 0 and I think using -1 should be reserved for a UNKNOWN flag.

NONE isn't a flag, it's a bitmask representing no flags being set. If we have something like `getFlags(): EnumSet<UserFlag>` then it makes no sense to return a `EnumSet.of(NONE)` rather than just an empty set. There is no reasonable way to use this. `UNKNOWN` is a placeholder constant used when we can't resolve the flag to an enum constant.

> I suggested here that we perhaps should have both the offset and the raw value? Would kind of make sense considering Discord provides the flags as their raw (total) value.

The offset gives a good and understandable code. Using raw values is confusing and can lead to bugs. You should simply use the offset to do `1 << offset` in a `getRaw()`. See [Permission](https://github.com/DV8FromTheWorld/JDA/blob/master/src/main/java/net/dv8tion/jda/api/Permission.java#L120). The [documentation](https://discordapp.com/developers/docs/resources/user#user-object-user-flags) also uses offsets for the same reason.

Andre601 commented:
I merged the suggested changes from @boomboompower in https://github.com/Andre601/JDA/pull/3 which also covered most, if not all requested changes by Minn

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -229,29 +229,29 @@ default String getEffectiveAvatarUrl()
-        VERIFIED_DEVELOPER(   131072, "Verified Developer");
-        
+    enum Flags {
+        UNKOWN(                   -1, "Unknown"),
```
```suggestion
        UNKNOWN(                   -1, "Unknown"),
```

Andre601 commented:
I think now should I implement this into the User object somehow?
I never done this before, so I'm unsure where I should alter stuff to implement this. A helping hand on this (i.e. guiding me into the right direction) would be appreciated.

MinnDevelopment commented:
The user objects are created in the `EntityBuilder` and the flags should be stored as an `int` in the `UserImpl` class. The getter is declared in the `User` interface and should return an `EnumSet`. The implementation will need the same getter plus one for the `int` itself, used to compare against in the `updateUser` method in `EntityBuilder` to fire an update event.

Andre601 commented:
Would the Permission enum be a good example for the Getter?

Andre601 commented:
I just noticed something.

In the Permission enum do you have this code:
```java
    @Nonnull
    public static EnumSet<Permission> getPermissions(long permissions)
    {
        if (permissions == 0)
            return EnumSet.noneOf(Permission.class);
        EnumSet<Permission> perms = EnumSet.noneOf(Permission.class);
        for (Permission perm : Permission.values())
        {
            if (perm != UNKNOWN && (permissions & perm.raw) == perm.raw)
                perms.add(perm);
        }
        return perms;
    }
```

Wouldn't it be somewhat better to have it like this?
```java
    @Nonnull
    public static EnumSet<Permission> getPermissions(long permissions)
    {
        EnumSet<Permission> perms = EnumSet.noneOf(Permission.class);
        if (permissions == 0)
            return perms;
        for (Permission perm : Permission.values())
        {
            if (perm != UNKNOWN && (permissions & perm.raw) == perm.raw)
                perms.add(perm);
        }
        return perms;
    }
```

MinnDevelopment commented:
Permissions are a good example for flags, yes.

Andre601 commented:
I now implemented a `getFlags()` of type `EnumSet<Flag>` to the User Object and updated the UserImpl accordingly.

Right now am I stuck with the following things:
- Not sure how I should set the flags (int) and enum set of it for UserImpl
- Not sure how to best return the flags (int)
- Best way to compare values in the EntityBuilder... Similar as the one used in `updateUser`? (`!Objects.equasls(Object, Object)`)?

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/UserImpl.java:
```diff
@@ -216,6 +231,12 @@ public UserImpl setFake(boolean fake)
         return this;
     }
+    
+    public UserImpl setFlags(int flags){
```
```suggestion
    public UserImpl setFlags(int flags)
    {
```

Andre601 commented:
I've added a `UserUpdateFlagsEvent` that would get triggered when the user's flags get updated.

Not sure if I made everything alright tho.

Andre601 commented on src/main/java/net/dv8tion/jda/api/Permission.java:
```diff
@@ -223,7 +223,7 @@ public static Permission getFromOffset(int offset)
     /**
      * A set of all {@link net.dv8tion.jda.api.Permission Permissions} that are specified by this raw long representation of
-     * permissions. The is best used with the getRaw methods in {@link net.dv8tion.jda.api.entities.Role Role} or
+     * permissions. This is best used with the getRaw methods in {@link net.dv8tion.jda.api.entities.Role Role} or
```
I wasn't sure if this was a typo (Probably was?) so I corrected this...
Or should it be `The method is ...`?

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,168 @@ default String getEffectiveAvatarUrl()
     JDA getJDA();
+
+    /**
+     * Returns the {@link net.dv8tion.jda.api.entities.User.Flag Flags} of this user.
```
Why don't you call it UserFlag? Could be a bit more readable

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,168 @@ default String getEffectiveAvatarUrl()
+        /**
+         * The binary offset of the flag.
+         * 
+         * @return The offset that represents this {@link Flag Flag}.
```
The [JDA Structure Guide](https://github.com/DV8FromTheWorld/JDA/wiki/6)-JDA-Structure-Guide) advises against documenting with imported names. I'm not sure if it doesn't apply here for some reason, but I'm mentioning it just in case

Andre601 replied:
I blame IntelliJ for auto-cutting the {@link} parameter as it had the full path before.

HydroPage90 replied:
Fair enough

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,168 @@ default String getEffectiveAvatarUrl()
     JDA getJDA();
+
+    /**
+     * Returns the {@link net.dv8tion.jda.api.entities.User.UserFlag Flags} of this user.
```
```suggestion
     * Returns the {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags} of this user.
```

Probably change these to accommodate the recent commit

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,168 @@ default String getEffectiveAvatarUrl()
+         * @return {@link net.dv8tion.jda.api.entities.User.UserFlag Flag} relating to the provided offset.
+         */
+        @Nonnull
+        public static UserFlag getFromOffset(int offset)
```
I wonder what a valid use-case for this would be. When would you work with the raw offsets in client-code? Isn't that exposing the implementation a bit too much?

What I'm saying is: if you're implementing a `UserFlag` enum to encapsulate the `public_flag` numbers, why would you allow the client code to still work with the numbers? Seems a bit off to me, I'd like to hear your thoughts

Andre601 replied:
Most of the code is based on the `Permission` enum, which also has a `getFromOffset` method for this.

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,168 @@ default String getEffectiveAvatarUrl()
+        @Nonnull
+        public static EnumSet<UserFlag> getFlags(int flags)
+        {
+            if(flags == 0)
+                return EnumSet.noneOf(UserFlag.class);
+            EnumSet<UserFlag> userFlags = EnumSet.noneOf(UserFlag.class);
+            for (UserFlag flag : UserFlag.values())
+            {
+                if (flag != UNKNOWN && (flags & flag.raw) == flag.raw)
+                    userFlags.add(flag);
+            }
+            return userFlags;
```
```suggestion
            final EnumSet<UserFlag> foundFlags = EnumSet.noneOf(UserFlag.class);
            
            if (flags == 0)
                return foundFlags; //empty
            
            for (UserFlag flag : values())
                if (flag != UNKNOWN && (flags & flag.raw) == flag.raw)
                    foundFlags.add(flag);
                    
            return foundFlags;
```

Not sure if this would help a bit, or if you want to keep the look of the method the exact same as in `Permissions`

Andre601 replied:
I also asked Minn why it is like that and not different (See [here](https://github.com/DV8FromTheWorld/JDA/pull/1271#issuecomment-619544061))
No idea why this comment was marked as off-topic.

I myself can't see a reason why `EnumSet.noneOf(Class)` is used twice as it could just be used once and modified later on...
I'll merge this suggestion and probably update the Permission class accordingly, unless Minn is against this.

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+        }
+
+        /**
+         * This is effectively the oposite of {@link #getFlags(int)}, this takes 1 or more {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}
```
```suggestion
         * This is effectively the opposite of {@link #getFlags(int)}, this takes 1 or more {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,77 @@
+        return getNewValue();
+    }
+
+    /**
+     * The new Set of {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     *
+     * @return The previous EnumSet of {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     */
```
```suggestion
    /**
     * @return The new EnumSet of {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}.
     */
```

You said "previous" before, even though this is the method for the new ones

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,77 @@
+        return previous == null ? null : User.UserFlag.getFlags(previous);
+    }
+
+    /**
+     * The new raw {@code Integer} representation of the set {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     * 
+     * @return The new {@code Integer} representation of the {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     */
```
```suggestion
    /**
     * @return The new {@code Integer} representation of the {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}.
     */
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java:
```diff
@@ -171,6 +171,7 @@ public void onUserUpdateActivityOrder(@Nonnull UserUpdateActivityOrderEvent even
     public void onUserTyping(@Nonnull UserTypingEvent event) {}
     public void onUserActivityStart(@Nonnull UserActivityStartEvent event) {}
     public void onUserActivityEnd(@Nonnull UserActivityEndEvent event) {}
+    public void onUserUpdateFlagsEvent(@Nonnull UserUpdateFlagsEvent event) {}
```
```suggestion
    public void onUserUpdateFlags(@Nonnull UserUpdateFlagsEvent event) {}
```

Follow the naming convention

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java:
```diff
@@ -459,6 +460,8 @@ else if (event instanceof UserActivityStartEvent)
         else if (event instanceof UserActivityEndEvent)
             onUserActivityEnd((UserActivityEndEvent) event);
+        else if (event instanceof UserUpdateFlagsEvent)
+            onUserUpdateFlagsEvent((UserUpdateFlagsEvent) event);
```
```suggestion
            onUserUpdateFlags((UserUpdateFlagsEvent) event);
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:
```diff
@@ -367,6 +371,15 @@ public void updateUser(UserImpl userObj, DataObject user)
                     userObj, oldAvatar));
         }
+        
+        if (!Objects.equals(oldFlags, newFlags))
```
Why not just simple `!=`? This is less readable

Andre601 replied:
Because the above one also uses Objects.equals, so I do this here too for consistency.

MinnDevelopment replied:
We use `Objects.equals` for things that can be null, int cannot be null so you should use `!=`.

Andre601 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         *         
+         * @return Unsigned long representing the provided flags.
+         */
+        public static int getRaw(@Nonnull UserFlag... flags){
+            int raw = 0;
+            for (UserFlag flag : flags)
+            {
+                if (flag != null && flag != UNKNOWN)
+                    raw |= flag.raw;
+            }
+            
+            return raw;
+        }
```
Long is from when this was a long and not an integer.
Also, this is - like many other things - based on the Permission enum, so I'll keep it like that for consistency.

HydroPage90 replied:
Understood

Andre601 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+        @Nonnull
+        public static UserFlag getFromOffset(int offset)
+        {
+            for (UserFlag flag : values())
+            {
+                if(flag.offset == offset)
+                    return flag;
+            }
+            return UNKNOWN;
```
See my previous post

Andre601 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         * @return Possibly-empty EnumSet of {@link UserFlag Flags}.
+         */
+        @Nonnull
+        public static EnumSet<UserFlag> getFlags(int flags)
```
It's int.
It was previously long, but I changed it, because Discord seems to send the flags as Integer and not Long.

HydroPage90 replied:
The documentation still says it's a long, that's what I was talking about

Andre601 commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    /**
+     * The previous raw {@code Integer} representation of the set {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     * 
+     * @return The previous {@code Integer} representation of the {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
```
I don't see where this is wrong.
I could rephrase it but don't see a reason as to why I should leave the comment itself away.

HydroPage90 replied:
The comment basically says "This returns the integer" and then you have a `@return` tag that literally says "This returns the integer". The comment itself feels redundant if the purpose can be more literally specified by other means, no?

Andre601 commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,77 @@
+        return getOldValue();
+    }
+
+    /**
+     * The previous Set of {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     * 
+     * @return The previous EnumSet of {@link net.dv8tion.jda.api.entities.User.UserFlag Flags}.
+     */
+    @Nullable
+    public EnumSet<User.UserFlag> getOlfFlagSet(){
+        return previous == null ? null : User.UserFlag.getFlags(previous);
+    }
```
Shouldn't it be UserFlags instead of Flags? 😉

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         * 
+         * @see java.util.EnumSet EnumSet
+         */
+        public static int getRaw(@Nonnull Collection<UserFlag> flags)
```
Again, the `@return` still claims `long`

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,69 @@
+     * @return Possibly-null EnumSet of previous {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}.
+     */
+    @Nullable
+    public EnumSet<User.UserFlag> getOlfFlagSet(){
```
```suggestion
    public EnumSet<User.UserFlag> getOldFlagSet(){
```

Still misspelled

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+        }
+
+        /**
+         * Gets the first {@link net.dv8tion.jda.api.entities.User.UserFlag Flag} relating to the provided offset.
+         * <br>If there is no {@link net.dv8tion.jda.api.entities.User.UserFlag Flag} that matches the provided offset,
+         * {@link net.dv8tion.jda.api.entities.User.UserFlag#UNKNOWN Flag.UNKNOWN} is returned.
+         * 
+         * @param  offset
+         *         The offset to match a {@link net.dv8tion.jda.api.entities.User.UserFlag Flag} to.
+         *         
+         * @return {@link net.dv8tion.jda.api.entities.User.UserFlag Flag} relating to the provided offset.
```
```suggestion
         * Gets the first {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} relating to the provided offset.
         * <br>If there is no {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} that matches the provided offset,
         * {@link net.dv8tion.jda.api.entities.User.UserFlag#UNKNOWN UserFlag.UNKNOWN} is returned.
         * 
         * @param  offset
         *         The offset to match a {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} to.
         *         
         * @return {@link net.dv8tion.jda.api.entities.User.UserFlag FlagFlag} relating to the provided offset.
```

This is tedious for me. Can you do it in your IDE?

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         * @param  flags
+         *         The array of flags of which to form into the raw long representation.
+         *         
+         * @return Unsigned long representing the provided flags.
```
```suggestion
         * @return Unsigned {@code int} representing the provided flags.
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         * @param  flags
+         *         The Collection of flags of which to form into the raw long representation.
+         *
+         * @return Unsigned long representing the provided flags.
```
```suggestion
         * @return Unsigned {@code int} representing the provided flags.
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         * flags.
+         * 
+         * @param  flags
+         *         The raw {@code long} representation if flags.
```
```suggestion
         *         The raw {@code int} representation if flags.
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+         * and returns the raw offset {@code int} representation of the flags.
+         * 
+         * @param  flags
+         *         The array of flags of which to form into the raw long representation.
```
```suggestion
         *         The array of flags of which to form into the raw {@code int} representation.
```

HydroPage90 commented:
Why is this still a draft? Do you have anything else in mind?

Andre601 commented:
It's a draft, because I'm not sure if everything was added, that has to be added.

HydroPage90 commented:
How do we make sure, then? Shall we review request Minn?

Andre601 commented:
That is probably the best option here.

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+        {
+            for (UserFlag flag : values())
+            {
+                if(flag.offset == offset)
```
```suggestion
                if (flag.offset == offset)
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,169 @@ default String getEffectiveAvatarUrl()
+            if (flags == 0)
+                return foundFlags; //empty
+            
+            for (UserFlag flag : values())
+                if (flag != UNKNOWN && (flags & flag.raw) == flag.raw)
+                    foundFlags.add(flag);
```
```suggestion
            for (UserFlag flag : values())
            {
                if (flag != UNKNOWN && (flags & flag.raw) == flag.raw)
                    foundFlags.add(flag);
            }
```

Didn't realize the convention for JDA was having braces on the loops

boomboompower commented:
Unmark the PR as a draft to recieve feedback ^

Andre601 commented:
Just because it's a draft doesn't mean it can't receive feedback. Your point is completely pointless.

DManstrator commented:
You can post feedback but when I see a MR on WiP / draft, I wouldn't give feedback since I don't want to say stuff on things which are not final yet. Maybe the author wants to do exactly what I have in my mind but they had no time to do that yet. So I can kinda understand it.

HydroPage90 commented:
@DManstrator I'd still comment, man. What you have on your mind could be a new idea we haven't discussed yet. And if we're working on it, there's no problem telling you. This is what PRs are for

DManstrator commented:
@HydroPage90 I'm literally too lazy for that. Worst case: You suggest things which are going to be implemented nevertheless -> waste of time.
Most people just want to see the final results and improve that. At least this is my attitude.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+        /**
+         * The readable name as used in the Discord Client.
+         * 
+         * @return The readable name of this {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag}.
```
Linking to the same class is redundant

Andre601 replied:
This should then be corrected in the Permission enum too.
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L129
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L142
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L203-L204
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L208
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L224
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L233
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L251
https://github.com/DV8FromTheWorld/JDA/blob/cc11f87957c2512ce33f84664df9df7ad43e2908/src/main/java/net/dv8tion/jda/api/Permission.java#L272

MinnDevelopment replied:
The permission enum is not the topic of this PR.

Andre601 replied:
I did remove any links to the same class in the UserFlag enum

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+         */
+        public static int getRaw(@Nonnull UserFlag... flags){
+            int raw = 0;
+            for (UserFlag flag : flags)
```
Add a `Checks.noneNull` for this, we don't want to throw NPE's if we can avoid it. Also add docs for the `IllegalArgumentException`.

Andre601 replied:
Added the check and the `@throws` about the IllegalArgumentException

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+         * @param  flags
+         *         The Collection of flags of which to form into the raw int representation.
+         *
+         * @return Unsigned int representing the provided flags.
```
Java doesn't support unsigned ints. Call it `bitmask`

Andre601 replied:
So `Unsigned int` to `Bitmask`, or `Unsigned bitmask`?

MinnDevelopment replied:
Just `bitmask` and yes best to refer to it as `bitmask` everywhere

Andre601 replied:
changed

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+        }
+
+        /**
+         * This is effectively the opposite of {@link #getFlags(int)}, this takes 1 or more {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}
```
```suggestion
         * This is effectively the opposite of {@link #getFlags(int)}. This takes a collection of UserFlags
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,69 @@
+package net.dv8tion.jda.api.events.user.update;
```
Missing copyright

Andre601 replied:
Added

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,69 @@
+ * member was updated and gives us the updated member object. In order to fire a specific event like this we
+ * need to have the old member cached to compare against.
+ */
+public class UserUpdateFlagsEvent extends GenericUserUpdateEvent<Integer>
```
I think this should be using the `EnumSet` instead of the int representation. Then you can add stuff like `getNewFlagsRaw()` here, etc.

Andre601 replied:
I changed it.
Please take a look and let me know if the changes are alright or need tweaking

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/UserImpl.java:
```diff
@@ -155,6 +158,18 @@ public boolean isFake()
+        return flagsSet;
+    }
+    
+    public int getFlagsRaw()
```
Add this to the interface. This can be useful for better compatibilty when discord adds new flags.

Andre601 replied:
Added

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/UserImpl.java:
```diff
@@ -43,6 +44,8 @@
     protected long privateChannel = 0L;
     protected boolean bot;
     protected boolean fake = false;
+    protected EnumSet<UserFlag> flagsSet;
```
This shouldn't be stored, just compute it when its needed.

Andre601 replied:
Was changed and is now computed in `getFlags()`

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/UserImpl.java:
```diff
@@ -155,6 +158,18 @@ public boolean isFake()
+    @Override
+    public EnumSet<UserFlag> getFlags()
+    {
+        return flagsSet;
```
You're returning a modifiable set here. This shouldn't be a problem if you would compute it inplace rather than ahead of time.

Andre601 replied:
Was changed to `UserFlag.getFlags(flags)`

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/UserImpl.java:
```diff
@@ -216,6 +231,13 @@ public UserImpl setFake(boolean fake)
+    public UserImpl setFlags(int flags)
+    {
+        this.flagsRaw = flags;
+        this.flagsSet = UserFlag.getFlags(flags);
```
Always converting these to an EnumSet is a bad idea, most user instances won't need this representation.

Andre601 replied:
removed the flagSet

Andre601 commented on src/main/java/net/dv8tion/jda/internal/entities/UserImpl.java:
```diff
@@ -43,6 +44,8 @@
     protected boolean bot;
     protected boolean fake = false;
+    protected EnumSet<Flag> flagsSet;
+    protected int flags;
```
Then how should I handle the Override of `getFlags()` which is a EnumSet?
Maybe return `UserFlag.getFlags(flags)`? Or shouldn't this be a thing in the User interface?
That would be against what you said (https://github.com/DV8FromTheWorld/JDA/pull/1271#issuecomment-619538928)

MinnDevelopment replied:
```java
EnumSet<UserFlag> getFlags()
{
  return UserFlag.getFlags(flags);
}
```

Andre601 replied:
Did it implement like that now.

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+        /**
+         * The readable name as used in the Discord Client.
+         * 
+         * @return The readable name of this {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag}.
```
```suggestion
```

HydroPage90 replied:
```suggestion
         * @return The readable name of this UserFlag
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+        }
+
+        /**
+         * Gets the first {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} relating to the provided offset.
+         * <br>If there is no {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} that matches the provided offset,
+         * {@link net.dv8tion.jda.api.entities.User.UserFlag#UNKNOWN UserFlag.UNKNOWN} is returned.
+         * 
+         * @param  offset
+         *         The offset to match a {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} to.
+         *         
+         * @return {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag} relating to the provided offset.
```
```suggestion
         * Gets the first UserFlag relating to the provided offset.
         * <br>If there is no UserFlag that matches the provided offset,
         * {@link net.dv8tion.jda.api.entities.User.UserFlag#UNKNOWN UNKNOWN} is returned.
         * 
         * @param  offset
         *         The offset to match a UserFlag to.
         *         
         * @return UserFlag relating to the provided offset.
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+        }
+        
+        /**
+         * A set of all {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags} that are specified by this raw int representation of
+         * flags.
+         * 
+         * @param  flags
+         *         The raw {@code int} representation if flags.
+         *         
+         * @return Possibly-empty EnumSet of {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}.
```
```suggestion
         * A set of all UserFlags that are specified by this raw int representation of
         * flags.
         * 
         * @param  flags
         *         The raw {@code int} representation if flags.
         *         
         * @return Possibly-empty EnumSet of UserFlag.
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+        }
+
+        /**
+         * This is effectively the opposite of {@link #getFlags(int)}, this takes 1 or more {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}
```
```suggestion
         * This is effectively the opposite of {@link #getFlags(int)}. This takes a collection of UserFlags
```

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+         * @return Unsigned int representing the provided flags.
+         */
+        public static int getRaw(@Nonnull UserFlag... flags){
+            int raw = 0;
+            for (UserFlag flag : flags)
```
```suggestion
            int raw = 0;
            
            Checks.noneNull(flags, "No flags should be null");
            for (UserFlag flag : flags)
```

Andre601 replied:
The second arg is the name of the argument provided and not a message.
I implemented it tho. ;)

HydroPage90 commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,171 @@ default String getEffectiveAvatarUrl()
+         * @param  flags
+         *         The array of flags of which to form into the raw int representation.
+         *         
+         * @return Unsigned int representing the provided flags.
```
```suggestion
         * @return bitmask representing the provided flags.
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -25,45 +40,45 @@
     @Nullable
-    public EnumSet<User.UserFlag> getOldFlagSet(){
-        return previous == null ? null : User.UserFlag.getFlags(previous);
+    public Integer getOldFlagsRaw(){
```
Change this to `int` and remove the `Nullable`

Andre601 replied:
I assume for both methods?
I changed it

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,183 @@ default String getEffectiveAvatarUrl()
+        /**
+         * Gets the first UserFlag relating to the provided offset.
+         * <br>If there is no UserFlag that matches the provided offset,
+         * {@link net.dv8tion.jda.api.entities.User.UserFlag#UNKNOWN UserFlag.UNKNOWN} is returned.
```
```suggestion
         * {@link #UNKNOWN} is returned.
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,183 @@ default String getEffectiveAvatarUrl()
+         * @return bitmask representing the provided flags.
+         * 
+         * @throws java.lang.IllegalArgumentException
+         *         When the provided UsefLags are null.
```
We usually put the `@throws` before the `@return`

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,183 @@ default String getEffectiveAvatarUrl()
+         */
+        public static int getRaw(@Nonnull Collection<UserFlag> flags)
+        {
+            Checks.notNull(flags, "Flag Collection");
```
Missing `@throws` for this exception

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,82 @@
+{
+    public static final String IDENTIFIER = "public_flags";
+    
+    public UserUpdateFlagsEvent(@Nonnull JDA api, long responseNumber, @Nonnull User user, EnumSet<User.UserFlag> oldFlags)
```
```suggestion
    public UserUpdateFlagsEvent(@Nonnull JDA api, long responseNumber, @Nonnull User user, @Nonnull EnumSet<User.UserFlag> oldFlags)
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @return The old {@code EnumSet<{@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag}>} representation of the User's flags.
+     */
+    @Nullable
```
```suggestion
    @Nonnull
```

Andre601 replied:
The change of this results in `getOldValue()` returning an error that it might evaluate as null while it is annotated as Nonnull

MinnDevelopment replied:
Ignore it.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @return The new {@code EnumSet<{@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag}>} representation of the User's flags.
+     */
+    @Nullable
```
```suggestion
    @Nonnull
```

Andre601 replied:
The change of this results in `getNewFlags()` returning an error that it might evaluate as null while it is annotated as Nonnull

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @return The old raw {@code Integer} value of the {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}.
+     */
+    public int getOldFlagsRaw(){
```
```suggestion
    public int getOldFlagsRaw()
    {
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,82 @@
+     * @return The new {@code EnumSet<{@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag}>} representation of the User's flags.
+     */
+    @Nullable
+    public EnumSet<User.UserFlag> getNewFlags(){
```
```suggestion
    public EnumSet<User.UserFlag> getNewFlags()
    {
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @return The old raw {@code Integer} value of the {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}.
+     */
+    public int getNewFlagsRaw(){
```
```suggestion
    public int getNewFlagsRaw()
    {
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/internal/entities/EntityBuilder.java:
```diff
@@ -367,6 +371,15 @@ public void updateUser(UserImpl userObj, DataObject user)
+            jda.handleEvent(
+                    new UserUpdateFlagsEvent(
+                        jda, responseNumber,
+                        userObj, oldFlags));
```
Doesn't this accept an `EnumSet` now?

Andre601 replied:
Should I change it to `User.UserFlag.getFlags(oldFlags)` or should I use a different approach?

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,186 @@ default String getEffectiveAvatarUrl()
+         *         The array of flags of which to form into the raw int representation.
+         *
+         * @throws java.lang.IllegalArgumentException
+         *         When the provided UserFLags are null.
```
```suggestion
         *         When the provided UserFlags are null.
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,186 @@ default String getEffectiveAvatarUrl()
+
+        /**
+         * This is effectively the opposite of {@link #getFlags(int)}, this takes 1 or more UserFlags
+         * and returns the raw offset {@code int} representation of the flags.
```
```suggestion
         * and returns the bitmask representation of the flags.
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,186 @@ default String getEffectiveAvatarUrl()
+
+        /**
+         * This is effectively the opposite of {@link #getFlags(int)}. This takes a collection of UserFlags
+         * and returns the raw offset {@code int} representation of the flags.
```
```suggestion
         * and returns the bitmask representation of the flags.
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/entities/User.java:
```diff
@@ -225,4 +228,186 @@ default String getEffectiveAvatarUrl()
+         * <br>Example: {@code getRaw(EnumSet.of(UserFlag.STAFF, UserFlag.HYPESQUAD))}
+         *
+         * @param  flags
+         *         The Collection of flags of which to form into the raw int representation.
```
```suggestion
         *         The flags to convert
```

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * @return The old {@code EnumSet<{@link net.dv8tion.jda.api.entities.User.UserFlag UserFlag}>} representation of the User's flags.
```
Missing description (same for the other methods).

Also:
```suggestion
     * @return {@link EnumSet} of the old {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}
```

Andre601 replied:
Should the Link for EnumSet not be the full path, rather than just EnumSet?

MinnDevelopment replied:
This way you don't need to make an alias for it. It's already imported so javadoc knows what to reference.

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/hooks/ListenerAdapter.java:
```diff
@@ -171,6 +171,7 @@ public void onUserUpdateActivityOrder(@Nonnull UserUpdateActivityOrderEvent even
     public void onUserTyping(@Nonnull UserTypingEvent event) {}
     public void onUserActivityStart(@Nonnull UserActivityStartEvent event) {}
     public void onUserActivityEnd(@Nonnull UserActivityEndEvent event) {}
+    public void onUserUpdateFlags(@Nonnull UserUpdateFlagsEvent event) {}
```
This should be grouped with the other update events above

Andre601 replied:
I don't quite understand what you mean exactly with that.

MinnDevelopment replied:
All the other `onUserUpdate` methods are grouped together a few lines above, this one should be in that group.

Andre601 replied:
Updated

MinnDevelopment commented on src/main/java/net/dv8tion/jda/api/events/user/update/UserUpdateFlagsEvent.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * @return {@link EnumSet} of the old {@link net.dv8tion.jda.api.entities.User.UserFlag UserFlags}
```
I think you missed my comment in the last review, methods need descriptions.

Andre601 replied:
Yeah, I missed it.
I did  had descriptions but iirc did @boomboompower (Or was it @HydroPage90?) suggest removing them to essentially prevent mentioning the same stuff twice.

MinnDevelopment replied:
That's a bad suggestion because the description is also used to show a short summary of the method in the list of methods. Redundancy is always good in the documentation. Better to say it twice than have people confused.

Andre601 replied:
I now added it.

Artuto commented:
update on this?

Andre601 commented:
> update on this?

Uhm..... It's ready?
I mean considering that Minn approved it, and that it has a set Milestone can we consider this done and the merge is the only thing waiting. So no need to hurry now.

HydroPage90 commented:
This is probably the largest conversation I'll have ever contributed to, lmao. Glad this seems to be done, and glad to have helped out a bit

Andre601 commented:
I would say it's kind of small if you remove all the review stuff.
